lib/accessmonitoring/monitor.go

```diff
@@ -0,0 +1,287 @@
+	Backend backend.Backend
+
+	// Events is the event monitor. This interface allows us to monitor access
+	// events.
+	Events types.Events
+
+	// FailFast mode indicates that the access monitor should fail fast instead
+	// of restarting indefinitely. Fail fast mode should only be used within a
+	// plugin environment. See https://github.com/gravitational/teleport/pull/30039
+	// for more details.
+	// TODO(bernardjkim): Investigate if fail fast mode is still required.
+	FailFast bool
+}
+
+// CheckAndSetDefaults checks and sets default config values.
+func (c *Config) CheckAndSetDefaults() error {
+	if c.Logger == nil {
+		c.Logger = slog.Default()
+	}
+	if c.Backend == nil {
+		return trace.BadParameter("backend: must be non-nil")
+	}
+	if c.Events == nil {
+		return trace.BadParameter("events: must be non-nil")
+	}
+	return nil
+}
+
+// AccessMonitor is an access monitoring service that monitors access events,
+// then executes the configured set of handlers for each event.
+type AccessMonitor struct {
+	cfg Config
+
+	// ruleHandlers contains the list of access monitoring rule event handlers.
+	ruleHandlers []EventHandler
+	// requestHandlers contains the list access request event handlers.
+	requestHandlers []EventHandler
+}
+
+// NewAccessMonitor returns a new access monitor.
+func NewAccessMonitor(cfg Config) (*AccessMonitor, error) {
+	if err := cfg.CheckAndSetDefaults(); err != nil {
+		return nil, trace.Wrap(err)
+	}
+	return &AccessMonitor{
+		cfg: cfg,
+	}, nil
+}
+
+// AddAccessMonitoringRuleHandler appends the access monitoring rule event handler.
+// These handlers are executed whenever a new access monitoring rule event is
+// observed.
+func (s *AccessMonitor) AddAccessMonitoringRuleHandler(handler EventHandler) {
+	s.ruleHandlers = append(s.ruleHandlers, handler)
+}
+
+// AddAccessRequestHandler appends the access request handler.
+// These handlers are executed whenever a new access request event is observed.
+func (s *AccessMonitor) AddAccessRequestHandler(handler EventHandler) {
+	s.requestHandlers = append(s.requestHandlers, handler)
+}
+
+// Run runs the access monitor. Restarts on failure.
+func (s *AccessMonitor) Run(ctx context.Context) error {
+	const retryJitter = 10 * time.Second
+	const retryInterval = 30 * time.Second
+	const lockTTL = time.Minute
+
+	waitWithJitter := retryutils.SeventhJitter(retryJitter)
+	for {
+		err := backend.RunWhileLocked(ctx, backend.RunWhileLockedConfig{
+			LockConfiguration: backend.LockConfiguration{
+				Backend:            s.cfg.Backend,
+				LockNameComponents: []string{componentName},
+				TTL:                lockTTL,
+				RetryInterval:      retryInterval,
+			},
+		}, s.run)
+		if err != nil {
+			s.cfg.Logger.ErrorContext(
+				ctx,
+				"Encountered a fatal error, it will restart after backoff.",
+				"component", componentName,
+				"error", err,
+				"restart_after", waitWithJitter,
+			)
+			if s.cfg.FailFast {
+				return trace.Wrap(err)
+			}
```

Comment:
If `FailFast` is enabled, it will not restart as the log line says.
Should we swap the log with the if block?

Reply:
Good catch. Updated in https://github.com/gravitational/teleport/pull/53769/commits/4277e42b0b9e33505a42d431f0cbff9857e67273